File: LeetCode/easy/Implement_Queue_using_Stacks.cxx
class MyQueue {
    std::stack<int> s_in, s_out;
    void reverse()
    {
        while (!s_in.empty())
        {
            s_out.push(s_in.top());
            s_in.pop();
        }        
    }
    
public:
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        s_in.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        if (s_out.empty())
            this->reverse();
        auto x = s_out.top();
        s_out.pop();
        return x;
    }
    
    /** Get the front element. */
    int peek() {
        if (s_out.empty())
            this->reverse();
        return s_out.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return s_in.empty() && s_out.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: LeetCode/easy/Implement_Queue_using_Stacks.py
class MyQueue:

    def __init__(self):
        self.s_in = []
        self.s_out = []

    def transverse(self):
        while len(self.s_in) > 0:
            self.s_out.append(self.s_in[-1])
            self.s_in.pop(-1)

    def push(self, x: int) -> None:
        self.s_in.append(x)
        

    def pop(self) -> int:
        if len(self.s_out) == 0:
            self.transverse()
        ret = self.s_out[-1]
        self.s_out.pop()
        return ret

    def peek(self) -> int:
        if len(self.s_out) == 0:
            self.transverse()
        return self.s_out[-1]

    def empty(self) -> bool:
        return len(self.s_in) == 0 and len(self.s_out) == 0
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
